src/layers/domain/core/product.py

```diff
@@ -1,98 +1,132 @@
-"""
-The collection of models required to create Products:
-
-OdsOrganisation -creates-> ProductTeam -creates-> Product
-                ∟ emits -> ProductTeamCreatedEvent
-"""
-from typing import Iterable, TypeVar
+from enum import Enum
 from uuid import UUID
 
-from domain.events.event import Event
+from domain.core.aggregate_root import AggregateRoot
+from domain.core.event import Event
+from pydantic import BaseModel, Field
+
+
+class ProductCreatedEvent(Event):
+    def __init__(self, id, name, type, product_team_id, ods_code, status):
+        self.id = id
+        self.name = name
+        self.type = type
+        self.product_team_id = product_team_id
+        self.ods_code = ods_code
+        self.status = status
+
+
+class ProductUpdatedEvent(Event):
+    def __init__(self, id, name, type):
+        self.id = id
+        self.name = name
+        self.type = type
+
+
+class ProductDeletedEvent(Event):
+    def __init__(self, id):
+        self.id = id
+
+
+class RelationshipAddedEvent(Event):
+    def __init__(self, id, target, type):
+        self.id = id
+        self.target = target
+        self.type = type
+
+
+class RelationshipRemovedEvent(Event):
+    def __init__(self, id, target):
+        self.id = id
+        self.target = target
+
+
+class ProductKeyAddedEvent(Event):
+    def __init__(self, id, key, type):
+        self.id = id
+        self.key = key
+        self.type = type
+
 
-from .entity import Entity
-from .questionnaire_entity import QuestionnaireEntity
-from .user import User
-from .validation import (
-    validate_ods_code,
-    validate_product_id_or_asid,
-    validate_type,
-    validate_uuid,
-)
+class ProductKeyRemovedEvent(Event):
+    def __init__(self, id, key):
+        self.id = id
+        self.key = key
 
-SetType = TypeVar("SetType")
 
+class RelationshipType(Enum):
```

Comment:
fyi in python 3.10+ you can do:

```suggestion
class RelationshipType(StrEnum):
    DEPENDENCY = auto()
    COMPONENT = auto()
    REVISION = auto()
```

which has some better behaviour

Reply:
Interesting.